presto-spi/src/main/java/io/prestosql/spi/block/MapHashTables.java

```diff
@@ -15,23 +15,128 @@
+    {
+        return INSTANCE_SIZE + sizeOf(hashTables);
+    }
+
+    boolean isBuilt()
+    {
+        return hashTables != null;
+    }
 
     /**
-     * This method assumes that {@code keyBlock} has no duplicated entries (in the specified range)
+     * Returns the raw hash tables, which must not be modified.
+     * @throws IllegalStateException if the hash tables have not been built
      */
-    static void buildHashTable(Block keyBlock, int keyOffset, int keyCount, MapType mapType, int[] outputHashTable, int hashTableOffset, int hashTableSize)
+    int[] get()
+    {
+        if (hashTables == null) {
+            throw new IllegalStateException("hashTables are not built");
+        }
+        return hashTables;
+    }
+
+    /**
+     * Returns the raw hash tables, if they have been built.  The raw hash tables must not be modified.
+     * @throws IllegalStateException if the hash tables have not been built
+     */
+    Optional<int[]> tryGet()
+    {
+        return Optional.ofNullable(hashTables);
+    }
+
+    synchronized void growHashTables(int newSize)
     {
+        int[] hashTables = this.hashTables;
+        if (hashTables == null) {
+            throw new IllegalStateException("hashTables size not set");
+        }
+        if (newSize < hashTables.length) {
+            throw new IllegalArgumentException("hashTables size does not match expectedEntryCount");
+        }
+        int[] newRawHashTables = Arrays.copyOf(hashTables, newSize);
+        Arrays.fill(newRawHashTables, hashTables.length, newSize, -1);
+        this.hashTables = newRawHashTables;
+    }
+
+    synchronized void buildAllHashTables(Block rawKeyBlock, int[] offsets, @Nullable boolean[] mapIsNull)
+    {
+        if (hashTables != null) {
+            return;
+        }
+
+        int[] hashTables = new int[rawKeyBlock.getPositionCount() * HASH_MULTIPLIER];
+        Arrays.fill(hashTables, -1);
+
+        int hashTableCount = offsets.length - 1;
+        for (int i = 0; i < hashTableCount; i++) {
+            int keyOffset = offsets[i];
+            int keyCount = offsets[i + 1] - keyOffset;
+            if (keyCount < 0) {
+                throw new IllegalArgumentException(format("Offset is not monotonically ascending. offsets[%s]=%s, offsets[%s]=%s", i, offsets[i], i + 1, offsets[i + 1]));
+            }
+            if (mapIsNull != null && mapIsNull[i] && keyCount != 0) {
+                throw new IllegalArgumentException("A null map must have zero entries");
+            }
+            buildHashTableInternal(rawKeyBlock, keyOffset, keyCount, hashTables);
+        }
+        this.hashTables = hashTables;
+    }
+
+    synchronized void buildHashTable(Block keyBlock, int keyOffset, int keyCount)
+    {
+        int[] hashTables = this.hashTables;
+        if (hashTables == null) {
+            throw new IllegalStateException("hashTables not set");
+        }
+
+        buildHashTableInternal(keyBlock, keyOffset, keyCount, hashTables);
```

Comment:
Is write to the fields of a volatile array inside a `synchronized` block guarantee visibility (without a volatile write)?
ie. Do we need `this.hashTables = hashTables` after the `buildHashTableInternal()`?

I did some search, but I can't draw a conclusion.
FYI: https://stackoverflow.com/questions/8827820/necessity-of-volatile-array-write-while-in-synchronized-block

Reply:
To clarify, the array is not volatile. What's volatile is the variable that points to the array. So writes to array elements are not guaranteed to be visible to readers of the array unless:
1. They synchronize on the same object.
2. They read a volatile variable that was written to after the elements of the array were written to.

Note that 2) just says that if the reader *sees* the updated value of the volatile, it's guaranteed to see the updated elements.

Reply:
I updated the code to write back to the variable after updaing